beforematch-security-and-privacy-assessment.md

```diff
@@ -0,0 +1,165 @@
+```html
+<style>
+  .hidden {
+    font-size: 3px;
+  }
+</style>
+<script>
+  onload = () => {
+    const hiddenElementContainer = document.createElement('div');
+    document.body.appendChild(hiddenElementContainer);
+
+    const label = document.createElement('div');
+    label.textContent = 'you are searching for:';
+    document.body.appendChild(label);
+
+    const resultElement = document.createElement('div');
+    document.body.appendChild(resultElement);
+
+    let result = '';
+
+    const nextResultElements = [];
+
+    const update = () => {
+      resultElement.textContent = result;
+      for (let i = 0; i < 128; i++) {
+        const element = nextResultElements[i];
+        element.textContent = result + String.fromCharCode(i);
+      }
+    };
+
+    for (let i = 0; i < 128; i++) {
+      const element = document.createElement('div');
+      element.className = 'hidden';
+      element.addEventListener('beforematch', () => {
+        result = element.textContent;
+        update();
+      });
+      hiddenElementContainer.appendChild(element);
+      nextResultElements.push(element);
+    }
+
+    update();
+  };
+</script>
+```
+
+### 2. What privacy attacks have been deemed out of scope (and why)?
+
+No attacks have been deemed out of scope.
+
+However, find-in-page attacks may already be possible due to the fact that pages can already intercept ctrl+f and already have the power to spoof the functionality and appearance of the browser's native find-in-page. In fact, doing this may be even easier and more reliable than using beforematch to extract the user's search text.
+
+For the ScrollToTextFragment case, you can already figure out what block-level element containing the search result was by looking at the element with the :target pseudo selector.
+
+### 3. What privacy mitigations have been implemented?
+
+No privacy mitigations have been implemented.
+
+For the find-in-page privacy flaw, I could imagine a mitigation where we would choose not to fire beforematch repeatedly in this case. In the example page above, the scroll offset would not be change in between firings of the event, which we could use as a heuristic to combat this case. However, I'm not sure that it couldn't be circumvented or if it would cause issues when trying to use the feature legitimately.
```

Comment:
```suggestion
For the find-in-page privacy case, I could imagine a mitigation where we would choose not to fire beforematch repeatedly in this case. In the example page above, the scroll offset would not be change in between firings of the event, which we could use as a heuristic to combat this case. However, I'm not sure that it couldn't be circumvented or if it would cause issues when trying to use the feature legitimately.
```